modules/services/window-managers/river.nix

```diff
@@ -0,0 +1,213 @@
+      else if isList value then
+        lists.flatten (map (x: toArgs path x) value)
+      else if value == null then
+        [ ]
+      else
+        [ stringValue ];
+    in finalValue;
+
+  # toCommand :: [string] -> attrs -> [string]
+  # Recursive function that converts an attrs to a list of commands that can be written to the
+  # config file.
+  toCommand = basePath: attrs:
+    concatLists (mapAttrsToList
+      (key: value: let path = basePath ++ [ key ]; in toArgs path value) attrs);
+
+in {
+  meta.maintainers = [ maintainers.GaetanLepage ];
+
+  options.wayland.windowManager.river = {
+    enable = mkEnableOption "river window manager";
+
+    package = mkPackageOption pkgs "river" {
+      nullable = true;
+      extraDescription = ''
+        Set to <code>null</code> to not add any River package to your path.
+        This should be done if you want to use the NixOS River module to install River.
+      '';
+    };
+
+    xwayland.enable = mkEnableOption "XWayland" // { default = true; };
+
+    systemd = {
+      enable = mkEnableOption null // {
+        default = true;
+        description = ''
+          Whether to enable {file}`river-session.target` on
+          river startup. This links to `graphical-session.target`.
+          Some important environment variables will be imported to systemd
+          and D-Bus user environment before reaching the target, including
+          - `DISPLAY`
+          - `WAYLAND_DISPLAY`
+          - `XDG_CURRENT_DESKTOP`
+          - `NIXOS_OZONE_WL`
+          - `XCURSOR_THEME`
+          - `XCURSOR_SIZE`
+        '';
+      };
+
+      variables = lib.mkOption {
+        type = with lib.types; listOf str;
+        default = [
+          "DISPLAY"
+          "WAYLAND_DISPLAY"
+          "XDG_CURRENT_DESKTOP"
+          "NIXOS_OZONE_WL"
+          "XCURSOR_THEME"
+          "XCURSOR_SIZE"
+        ];
+        example = [ "-all" ];
+        description = ''
+          Environment variables to be imported in the systemd & D-Bus user
+          environment.
+        '';
+      };
+
+      extraCommands = lib.mkOption {
+        type = with lib.types; listOf str;
+        default = [
+          "systemctl --user stop river-session.target"
+          "systemctl --user start river-session.target"
+        ];
+        description = "Extra commands to be run after D-Bus activation.";
+      };
+    };
+
+    extraSessionVariables = mkOption {
+      type = types.attrs;
+      default = { };
+      description = "Extra session variables set when running the compositor.";
+      example = { MOZ_ENABLE_WAYLAND = "1"; };
+    };
+
+    settings = mkOption {
+      type = with types;
+        let
+          valueType = nullOr (oneOf [
+            bool
+            int
+            float
+            str
+            path
+            (attrsOf valueType)
+            (listOf valueType)
+          ]) // {
+            description = "River configuration value";
+          };
+        in valueType;
```

Comment:
In the end, I haven't felt that it was necessary to use a `types.submodule` with `freeformType`.
Indeed, I haven't found necessary to explicitly declare any option in this case.

By the way, this is the same thing that is done in the [hyprland module](https://github.com/nix-community/home-manager/blob/3d6791b3897b526c82920a2ab5f61d71985b3cf8/modules/services/window-managers/hyprland.nix#L113-L127).